### PR TITLE
fix(logger): logger throws TypeError when log item has BigInt value

### PR DESCRIPTION
**Pull request by shdq**
## Description of your changes

The replacer function used for `JSON.stringify` converts `BigInt` values to strings.

Added tests and removed external dependency for `removeEmptyKeys()` function.

### Related issues, RFCs

**Issue number:** #1117 

### PR status

***Is this ready for review?:*** YES
***Is it a breaking change?:*** NO

## Checklist

- [x] [My changes meet the tenets criteria](https://awslabs.github.io/aws-lambda-powertools-typescript/#tenets)
- [x] I have performed a *self-review* of my own code
- [x] I have *commented* my code where necessary, particularly in areas that should be flagged with a TODO, or hard-to-understand areas
- [ ] I have made corresponding changes to the *documentation*
- [ ] I have made corresponding changes to the *examples*
- [x] My changes generate *no new warnings*
- [x] The *code coverage* hasn't decreased
- [x] I have *added tests* that prove my change is effective and works
- [x] New and existing *unit tests pass* locally and in Github Actions
- [x] Any *dependent changes have been merged and published*
- [x] The PR title follows the [conventional commit semantics](https://github.com/awslabs/aws-lambda-powertools-typescript/blob/main/.github/semantic.yml#L2)

### Breaking change checklist

- [ ] I have documented the migration process
- [ ] I have added, implemented necessary warnings (if it can live side by side)

---

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: packages/logger/src/log/LogItem.ts [hunk 1]
```diff
@@ -31,7 +30,14 @@ class LogItem implements LogItemInterface {
...
-    return pickBy(attributes, (value) => value !== undefined && value !== '' && value !== null);
+    const newAttributes: LogAttributes = {};
+    for (const key in attributes) {
+      if (attributes[key] !== undefined && attributes[key] !== '' && attributes[key] !== null) {
+        newAttributes[key] = attributes[key];
+      }
+    }
+    
+    return newAttributes;
   }
```
Comment (shdq) [inline packages/logger/src/log/LogItem.ts:41]: Firstly, I wrote a declarative one-liner:
```
public removeEmptyKeys(attributes: LogAttributes): LogAttributes {
    return Object.fromEntries(Object.entries(attributes).filter(([, value]) => value !== undefined && value !== '' && value !== null));
}
```
But for the sake of readability, I stick to the verbose version.

Comment (shdq) [reply packages/logger/src/log/LogItem.ts:41]: I don't know if nested objects could be in the `logItem`, but it wasn't and isn't supported by this function.

Comment (dreamorosi) [reply packages/logger/src/log/LogItem.ts:41]: Could you give an example of what do you mean by nested objects?

**Comment by dreamorosi** [comment]
I see that there's no `package-lock.json` in the file diff, if you have done it differently, could you please remove the dependency by running `npm remove lodash.pickby -w packages/logger`?

This should make the necessary updates to the lock file as well as removing it from the `package.json` of the package. Note that in order to successfully be able to run the command you might have to temporarily revert the change to `package.json`, otherwise npm will think the two files are out of sync.

Could you please also remove `@types/lodash.pickby` from `devDependencies` in the same package while at it?

File: packages/logger/src/log/LogItem.ts (hunk 1 above)
Comment (shdq) [reply packages/logger/src/log/LogItem.ts:41]: `logger.info("my message", { value: 42, nested: { value: "nested value", emptyValue: '', undefinedValue: undefined, nullValue: null } });`

**Comment by dreamorosi** [comment]
Also, I see that [the branch you're working on](https://github.com/shdq/aws-lambda-powertools-typescript/tree/1117-bug-logger-cant-serialize-bigint-values) is behind `main`.

Normally this would be fine, but in this case I would strongly advise to rebase/bring it in line with `main`. 

The reason why I say this is that as part of the migration to Node 18 we have updated all the dependencies and added the runtime to the unit & integration tests. For the integration tests to pass we need the CDK version that is currently present on `main` but not in your branch, which [causes the integration tests to fail](https://github.com/awslabs/aws-lambda-powertools-typescript/actions/runs/3794247482).

File: packages/logger/src/Logger.ts [hunk 2]
```diff
@@ -568,6 +568,37 @@ class Logger extends Utility implements ClassThatLogs {
...
+  /**
+   * When the data added in the log item contains object references or BigInt values,
+   * `JSON.stringify()` can't handle them and instead throws errors:
+   * `TypeError: cyclic object value` or `TypeError: Do not know how to serialize a BigInt`.
+   * To mitigate these issues, this method will find and remove all cyclic references and convert BigInt values to strings.
+   *
+   * @see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify#exceptions
+   * @private
+   */
+  private getReplacer(): (key: string, value: LogAttributes | Error | bigint) => void {
```
Comment (dreamorosi) [inline packages/logger/src/Logger.ts:580]: I like the new name, thanks for making the change

File: packages/logger/src/log/LogItem.ts (hunk 1 above)
Comment (dreamorosi) [reply packages/logger/src/log/LogItem.ts:41]: I think we need to support nested objects though. 

Here's an example of logs emitted by Logger `v1.5.0` (latest):
Screenshot 2022-12-28 at 14 40 48
 
from this code:
```ts
import { Logger, injectLambdaContext } from "@aws-lambda-powertools/logger";
import middy from "@middy/core";

const logger = new Logger({});

export const handler = middy(async () => {
  logger.info("my message", {
    value: 42,
    nested: {
      value: "nested value",
      emptyValue: "",
      undefinedValue: undefined,
      nullValue: null,
    },
  });
}).use(injectLambdaContext(logger, { logEvent: true }));
```

this was the event payload passed:
```json
{
  "key1": "value1",
  "key2": "value2",
  "key3": "value3",
  "nested": {
      "some": "key",
      "other": 1
  }
}
```

We need to support nested objects both for the `logEvent` functionality (also present in the unit tests) and arbitrary objects

Comment (dreamorosi) [reply packages/logger/src/log/LogItem.ts:41]: Oh after re-reading I realize I might have misunderstood your comment.

Were you referring to supporting the removal of `undefined`, `null`, etc. from nested objects?

Comment (shdq) [reply packages/logger/src/log/LogItem.ts:41]: > Oh after re-reading I realize I might have misunderstood your comment.
> 
> Were you referring to supporting the removal of `undefined`, `null`, etc. from nested objects?

yes, it's just go through the first level

Comment (dreamorosi) [reply packages/logger/src/log/LogItem.ts:41]: I see, apologies for the misunderstanding!

I think that at the moment we should strive for maintaining the same exact functionality level as the current version without breaking changes. Which based on the test I did above, seems to at least remove `undefined` values.

Then later on, if there's demand we can explore extending the behavior to empty strings and/or `null`.

Comment (shdq) [reply packages/logger/src/log/LogItem.ts:41]: `undefined` values are filtered somewhere else 🙃

**Comment by dreamorosi** [comment]
Thank you for syncing the branch & updating the dependencies & lock file.

I can confirm that the integration tests are passing: https://github.com/awslabs/aws-lambda-powertools-typescript/actions/runs/3794471678

File: packages/logger/src/log/LogItem.ts (hunk 1 above)
Comment (dreamorosi) [reply packages/logger/src/log/LogItem.ts:41]: And I definitely support/agree with the choice of using a more verbose version for the sake of readability.

**Review by dreamorosi — APPROVED** [review]
Thanks a lot @shdq for picking this up and for the quick PR.

I appreciate the choices made throughout.

**Comment by shdq** [comment]
> Thanks a lot @shdq for picking this up and for the quick PR.
> 
> I appreciate the choices made throughout.

I'm happy to help! Thanks for the fast feedback loop @dreamorosi!

**Pull request merged**